build/legacy-mongo-shell/README.md

```diff
@@ -0,0 +1,220 @@
+
+For example, to check if the `findAndModify` command failed with the error code `ImmutableField` you would do the following:
+
+```js
+> const res = db.runCommand({findAndModify: "foo", query: {}, update: {$set: {_id: 1}}});
+> assert.commandFailedWithCode(res, ErrorCodes.ImmutableField);
+```
+
+It is not always necessary use the `db.runCommand()` helper as some write methods wrap a `WriteResult` which the helpers can parse. For example, `insert`, `update`, and `remove` will all return a `WriteResult` so the function can parse the result and look for a `writeError`.
+
+```js
+> assert.commandWorked(db.foo.insert({a: 1}));
+WriteResult({ "nInserted" : 1 })
+> const res = db.foo.insert({a: 1});
+> res.hasWriteError();
+false
+```
+
+ The `find()` method will return an object whose properties will not be parsed by the various assert functions:
+
+```js
+> const res = db.foo.find({a: 1});
+> res._filter
+{ "a" : 1 }
+> res._db
+test
+> res._batchSize
+0
+```
+
+But a `runCommand()` result will return the correct object that can be parsed:
+
+```js
+> // use a runCommand instead
+> const res = db.runCommand({find: "foo", filter: {a: 1}});
+> res.ok
+1
+> res._commandObj
+{
+	"find" : "foo",
+	"filter" : {
+		"a" : 1
+	},
+	"lsid" : {
+		"id" : UUID("cb64879f-6f69-4656-a38e-ce6dbe3ccebc")
+	}
+}
+> assert.commandFailed(res);
+uncaught exception: Error: command worked when it should have failed: {
+	"cursor" : {
+		"firstBatch" : [
+			{
+				"_id" : ObjectId("6458ec7adb858f891c0b8c68"),
+				"a" : 1
+			},
+			{
+				"_id" : ObjectId("6458ec9ddb858f891c0b8c6a"),
+				"a" : 1
+			}
+		],
+		"id" : NumberLong(0),
+		"ns" : "test.foo"
+	},
+	"ok" : 1
+} :
+_getErrorWithCode@src/mongo/shell/utils.js:24:13
+doassert@src/mongo/shell/assert.js:18:14
+_assertCommandFailed@src/mongo/shell/assert.js:819:25
+assert.commandFailed@src/mongo/shell/assert.js:877:16
+@(shell):1:8
+```
+
+More useful functions:
+
+`assert.eq(a, b, msg)`
+
+throws if two values are not equal (tested without strict equality).
+
+`assert.isnull(what, msg)`
+
+throws if `what` is not null.
+
+`assert.commandFailed(res, msg)`
+
+throws if the command did not fail.
+
+`assert.commandWorked(res)`
+
+throws if the result contained an error.
+
+`assert.commandFailedWithCode(res, expectedCode, msg)`
+
+throws if the result did not contain the expected code.
+
+`assert.sameMembers(aArr, bArr, msg, compareFn = _isDocEq)`
+
+throws if the two arrays do not have the same members, in any order. By default, nested arrays must have the same order to be considered equal. Optionally accepts a compareFn to compare values instead of using docEq.
+
+`assert.writeOK(res, msg, {ignoreWriteConcernErrors} = {})`
```

Comment:
What is that syntax: `{ignoreWriteConcernErrors} = {}` ?

Reply:
https://masteringjs.io/tutorials/fundamentals/parameters – I think that's that a part of function definition, but not how it is supposed to be used, no?

Reply:
What you said makes sense. Yes I believe they are called named parameters and provide two benefits, 1. they are named so there's some improved readability and 2. they can be ignored AFAIK.

For instance see [here](https://github.com/mongodb/mongo/blob/master/src/mongo/shell/assert.js#L891), it checks if `ignoreWriteConcernErrors` is false and that the result had any write concern errors and then prints an error message with [failed write concern](https://github.com/mongodb/mongo/blob/master/src/mongo/shell/assert.js#LL892C27-L892C59).

Reply:
But I believe you can set them and pass them as arguments to the function.

```js
var f = function({optional} = {}) {
  if (!optional) {
    console.log("optional not provided");
  }
}
f()
optional not provided
f({optional: true});
undefined
```

Reply:
In short you can simply ignore them.

Reply:
Do you use it like `assert.writeOK(res, msg, {ignoreWriteConcernErrors: true})` and ignore it like  `assert.writeOK(res, msg)` ?

Reply:
I believe so, most the usage simply ignores it and just uses a message upon an error. You can also omit a message and let the assertion build the message for your, see [here](https://github.com/mongodb/mongo/blob/master/src/mongo/shell/assert.js#LL892C27-L892C59).

Write concern is specific to replication so let's not worry about using that parameter. We can just ignore it.

Reply:
If that's the case, for documentation I think using usage `assert.writeOK(res, msg, {ignoreWriteConcernErrors: true})` instead of syntax is `assert.writeOK(res, msg, {ignoreWriteConcernErrors} = {})` is easier to copy and paste for those not familiar us js :hugs:

Reply:
I think you are right, however to not diverge from the source code I think I should probably modify our documentation not the usage or signature perhaps? I mean highlighting the fact that we can ignore named parameters.

@chilagrow what do you think?

Reply:
I think we could omit it completely given that we don't use write concerns

Reply:
Sounds good, let me make that change.

Reply:
It looks like the syntax is removed from the article, can we resolve it to merge it?